app/controllers/concerns/foreman/controller/session.rb

```diff
@@ -3,6 +3,14 @@ module Foreman::Controller::Session
 
   def session_expiry
     return if ignore_api_request?
+    # if session[:sso_method] == "SSO::OpenidConnect", and access token has expired
+      # if refresh token has expired
+         # expire_sessiom
+      # else
+        # refresh_access_token
+      # end
+    # end
+
```

Comment:
Please see this without the hound errors :( I want to implement this in another PR but have put it here so that we have a better idea of what I want to achieve.

Reply:
Here, I  will handle the refresh token part, if refresh token expires then we must call he `expire_session` method and redirect to the SSO login page instead of Foreman login page.